src/core/lib/event_engine/query_extensions.h

```diff
@@ -0,0 +1,47 @@
+#ifndef GRPC_SRC_CORE_LIB_EVENT_ENGINE_QUERY_EXTENSIONS_H_
+#define GRPC_SRC_CORE_LIB_EVENT_ENGINE_QUERY_EXTENSIONS_H_
+
+#include <grpc/support/port_platform.h>
+
+#include <string_view>
+
+#include <grpc/event_engine/event_engine.h>
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// A helper class to derive from some set of base classes and export
+// QueryExtension for them all.
+// Endpoint implementations which need to support different extensions just need
+// to derive from EndpointExtensions class.
+template <typename... Exports>
+class EndpointExtensions : public EventEngine::Endpoint, public Exports... {
+ protected:
+  void* BasicQueryExtension(std::string_view id) {
+    return BasicQueryExtensionRecursion<Exports...>::Query(id, this);
+  }
+
+ private:
+  template <typename... Es>
+  struct BasicQueryExtensionRecursion;
+
+  template <typename E, typename... Es>
```

Comment:
Some compilers are going to have a hard time with an inner partial template specialization... suggest making an `grpc_event_engine::experimental::endpoint_detail` namespace and rewriting like so:

```
template <typename Querying, typename... Es>
struct BasicQueryExtensionRecursion;

template <typename Querying, typename E, typename... Es>
struct BasicQueryExtensionRecursion<Querying, E, Es...> {
  static void* Query(std::string_view id, Querying* p) {
    if (id == E::EndpointExtensionName()) return static_cast<E*>(p);
    return BasicQueryExtensionRecursion<Es...>::Query(id, p);
  }
};

template <typename Querying>
struct BasicQueryExtensionRecursion<Querying> {
  static void* Query(std::string_view, Querying*) {
    return nullptr;
  }
};
```

Reply:
Making line 21:

```
return endpoint_detail::BasicQueryExtensionRecursion<EndpointExtensions, Exports...>::Query(id, this);
```

Reply:
I didn't know that. Updated it now.

Reply:
I think we should avoid templates here. Putting this type in the public API invites users to write their own. How will this work if gRPC is compiled separately from their application?

Reply:
Per our offline discussion, moved the template as a helper function to src/core/lib/event_engine/query_extensions.h